crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -49,6 +54,215 @@ pub trait WorldQuery {
+///     for mut health in health_query.iter_mut() {
+///         health.damage(1.0);
+///         println!("Total (mut): {}", health.total());
+///     }
+/// }
+/// # my_system.system();
+/// ```
+///
+/// If you want to use derive macros with read-only query variants, you need to pass them with
+/// using the `read_only_derive` attribute. When `Fetch` macro generates an additional struct
+/// for a mutable query, it doesn't automatically inherit the same derives. Since derive macros
+/// can't access information about other derives, they need to be passed manually with the
+/// `read_only_derive` attribute.
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// use bevy_ecs::query::{Fetch, ReadOnlyFetch, WorldQuery};
+///
+/// #[derive(Component, Debug)]
+/// struct Foo;
+///
+/// #[derive(Fetch, Debug)]
+/// #[mutable]
+/// #[read_only_derive(Debug)]
+/// struct FooQuery<'w> {
+///     foo: &'w Foo,
+/// }
+///
+/// fn assert_debug<T: std::fmt::Debug>() {}
+///
+/// assert_debug::<FooQuery>();
+/// assert_debug::<FooQueryReadOnly>();
+/// ```
+///
+/// **Note:** if you omit the `mutable` attribute for a query that doesn't implement
+/// `ReadOnlyFetch`, compilation will fail. We insert static checks as in the example above for
+/// every query component and a nested query.
+/// (The checks neither affect the runtime, nor pollute your local namespace.)
+///
+/// ```compile_fail
+/// # use bevy_ecs::prelude::*;
+/// use bevy_ecs::query::{Fetch, ReadOnlyFetch, WorldQuery};
+///
+/// #[derive(Component)]
+/// struct Foo;
+/// #[derive(Component)]
+/// struct Bar;
+///
+/// #[derive(Fetch)]
+/// struct FooQuery<'w> {
+///     foo: &'w Foo,
+///     bar_query: BarQuery<'w>,
+/// }
+///
+/// #[derive(Fetch)]
+/// #[mutable]
+/// struct BarQuery<'w> {
+///     bar: Mut<'w, Bar>,
+/// }
+/// ```
+///
+/// ## Nested queries
+///
+/// Using nested queries enable the composition pattern, which makes it possible to re-use other
+/// query types. All types that implement [`WorldQuery`] (including the ones that use this derive
+/// macro) are supported.
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// use bevy_ecs::query::Fetch;
+///
+/// #[derive(Component)]
+/// struct Foo;
+/// #[derive(Component)]
+/// struct Bar;
+/// #[derive(Component)]
+/// struct OptionalFoo;
+/// #[derive(Component)]
+/// struct OptionalBar;
+///
+/// #[derive(Fetch)]
+/// struct MyQuery<'w> {
+///     foo: FooQuery<'w>,
+///     bar: (&'w Bar, Option<&'w OptionalBar>)
+/// }
+///
+/// #[derive(Fetch)]
+/// struct FooQuery<'w> {
+///     foo: &'w Foo,
+///     optional_foo: Option<&'w OptionalFoo>,
+/// }
+///
+/// // You can also compose derived queries with regular ones in tuples.
+/// fn my_system(query: Query<(&Foo, MyQuery, FooQuery)>) {
+///     for (foo, my_query, foo_query) in query.iter() {
+///         foo; my_query; foo_query;
+///     }
+/// }
+///
+/// # my_system.system();
+/// ```
 pub trait Fetch<'world, 'state>: Sized {
     type Item;
```

Comment:
This should prevent users forgeting to implement `FetchedItem` and thus make `WorldQuerys` that don't support the derive.
```suggestion
    type Item: FetchedItem;
```

Reply:
Yeah, I tried to do this, but unfortunately, it won't work for filters, as their `Item` is `bool`. And `bool` can mean absolutely any filter...

Reply:
Can we `impl FetchedItem for bool`?

Reply:
@alice-i-cecile I'm afraid we can't. `<<With<T> as WorldQuery>::Fetch as Fetch>::Item` corresponds to `bool`. So does `Without<T>`. As it's not a one-to-one relation for filters, I think there's no way to implement it.

Reply:
Maybe we'll be able to do it one day if we ban using filters as regular queries' items.

Reply:
Ok, I guess we can do it like so: https://github.com/bevyengine/bevy/pull/2713/commits/e65d2740ccb8376141e6f73301aa8ad6ccc2dc88

It's obviously useless for determining a filter type but allows us to add `FetchedItem` bound to `Fetch::Item`.